olive/passes/onnx/onnx_dag.py

```diff
@@ -0,0 +1,492 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+import logging
+from collections import defaultdict
+from enum import Enum
+from pathlib import Path
+from typing import TYPE_CHECKING, Dict, List, Set, Union
+
+import onnx
+from onnx import AttributeProto, GraphProto, NodeProto, TensorProto, ValueInfoProto
+
+from olive.common.config_utils import ConfigBase
+from olive.common.pydantic_v1 import Field
+
+if TYPE_CHECKING:
+    from onnx import ModelProto
+
+logger = logging.getLogger(__name__)
+
+
+class SpecialInput(str, Enum):
+    """Special inputs for ONNX nodes."""
+
+    INPUT = "__input__"  # user input
+    INITIALIZER = "__initializer__"  # constant initializer
+
+
+class SpecialOutput(str, Enum):
+    """Special outputs for ONNX nodes."""
+
+    OUTPUT = "__output__"  # model output
+
+
+class OnnxNode(ConfigBase):
+    """ONNX node."""
+
+    op_type: str
+    inputs: List[str]
+    outputs: List[str]
+    graph_idx: int
+    proto: NodeProto  # reference to the node in the model graph
+
+
+class OnnxIO(ConfigBase):
```

Comment:
>ConfigBase [](http://example.com/codeflow?start=13&length=10)

need the two classes be configured using pydantic?

Reply:
ConfigBase behaves like a dataclass. So, I am using it since I like the syntax and it's behaviors.

Reply:
could we directly use dataclass?

Reply:
as discussed offline. I will keep it as configbase but added comment about serialization to avoid potential confusion.